src/xdp/offloadqeo.c

```diff
@@ -0,0 +1,408 @@
+    BOOLEAN RundownAcquired = FALSE;
+    BOOLEAN LockHeld = FALSE;
+
+    TraceEnter(TRACE_CORE, "Interface=%p", InterfaceObject);
+
+    InitializeListHead(&QeoParams.Connections);
+    QeoParams.ConnectionCount = 0;
+
+    QeoSettings =
+        &XdpIfGetOffloadIfSettings(
+            InterfaceObject->IfSetHandle, InterfaceObject->InterfaceOffloadHandle)->Qeo;
+
+    if (InputBufferLength == 0 || OutputBufferLength != InputBufferLength) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    //
+    // Acquire an interface offload rundown reference to ensure offload cleanup
+    // waits until all QEO pre- and post-processing has completed.
+    //
+    if (XdpIfAcquireOffloadRundown(InterfaceObject->IfSetHandle)) {
+        RundownAcquired = TRUE;
+    } else {
+        Status = STATUS_DEVICE_NOT_READY;
+        goto Exit;
+    }
+
+    RtlAcquirePushLockExclusive(&QeoSettings->Lock);
+    LockHeld = TRUE;
+
+    while (InputBufferLength > 0) {
+        XDP_OFFLOAD_QEO_CONNECTION *OffloadConnection = NULL;
+
+        //
+        // Validate input.
+        //
+
+        if (InputBufferLength < sizeof(ConnectionsIn->Header) ||
+            InputBufferLength < ConnectionsIn->Header.Size) {
+            TraceError(
+                TRACE_CORE,
+                "Interface=%p Input buffer length too small InputBufferLength=%u",
+                InterfaceObject, InputBufferLength);
+            Status = STATUS_INVALID_PARAMETER;
+            goto Exit;
+        }
+
+        if (ConnectionsIn->Header.Revision != XDP_QUIC_CONNECTION_REVISION_1 ||
+            ConnectionsIn->Header.Size < XDP_SIZEOF_QUIC_CONNECTION_REVISION_1) {
+            TraceError(
+                TRACE_CORE, "Interface=%p Unsupported revision Revision=%u Size=%u",
+                InterfaceObject, ConnectionsIn->Header.Revision, ConnectionsIn->Header.Size);
+            Status = STATUS_INVALID_PARAMETER;
+            goto Exit;
+        }
+
+        if ((UINT32)ConnectionsIn->Operation > (UINT32)XDP_QUIC_OPERATION_REMOVE ||
+            (UINT32)ConnectionsIn->Direction > (UINT32)XDP_QUIC_DIRECTION_RECEIVE ||
+            (UINT32)ConnectionsIn->DecryptFailureAction > (UINT32)XDP_QUIC_DECRYPT_FAILURE_ACTION_CONTINUE ||
+            (UINT32)ConnectionsIn->CipherType > (UINT32)XDP_QUIC_CIPHER_TYPE_AEAD_AES_128_CCM ||
+            (UINT32)ConnectionsIn->AddressFamily > (UINT32)XDP_QUIC_ADDRESS_FAMILY_INET6 ||
+            ConnectionsIn->ConnectionIdLength > sizeof(ConnectionsIn->ConnectionId)) {
+            Status = STATUS_INVALID_PARAMETER;
+            goto Exit;
+        }
+
+        switch (ConnectionsIn->Operation) {
+        case XDP_QUIC_OPERATION_ADD:
+            if (XdpOffloadQeoFindOffloadedConnection(QeoSettings, ConnectionsIn)) {
+                Status = STATUS_DUPLICATE_OBJECTID;
+                goto Exit;
+            }
+
+            OffloadConnection =
+                ExAllocatePoolZero(
```

Comment:
Should we have a lookaside list for these?

Reply:
Yeah, probably eventually. Let's measure later.